### DISCUSSION TITLE
How to subscribe to events emitted from a plugin?

### DISCUSSION THREAD

**Discussion by michaelfung** [mojolicious/mojo#1890] [Q&A] [answered]
I want to emit some event, say **foo**, inside a plugin:

```
# inside a plugin module:
sub register ($self, $app, $conf) {
  ...
   $self->emit(foo => $event_data);

```

How can I subscribe to the  **foo** event at some other parts of the app?

**Comment by guest20** [comment, marked as answer]
You want the plugin to register a helper that makes `$self` available. If it's not your plugin you can just sub-class it and add the helper yourself.

```perl
# inside a plugin module:
sub register ($self, $app, $conf) {
  ...
   $app->helper(pluginName => sub { $self }); 
```
```perl
#elsewhere: 
$c->pluginName->on(...)
```

... it might be a bit late if the plugin emits an event during register though

**Reply by michaelfung** [reply]
I have thought about this but I am hoping for something more neat like:

```
my $myplugin = $plugin_manager->get_instance('MyPlugin');
$myplugin->on(foo => sub { ... });
```

My use case is: a MQTT plugin emit some topic message events, then I can add other plugins later when needed to inteface with Thingspeak, IFTTT, Matrix ...etc.

I believe we have to live with this solution for now. Thanks.
